New version of PoldiAutoCorrelation algorithm (v6) post detector upgrade

RichardWaiteSTFC opened the pull request:
### Description of work
PR includes:

- New version (v6) of algorithm to  PoldiAutoCorrelation algorithm for POLDI post detector upgrade (see new docs for details).
- Added silicon data and calibrated IDF in unit test data
- Unit test for PoldiAutoCorrelation algorithm (v6)
- Added `poldi_utils` file with helper functions that are common to the auto-correlation and later 2D workspace fitting
- Function in `poldi_utils` to simulate 2D workspace `simulate_2d_data`

POLDI scientists wanted the algorithm to be in python and to have same name as previous algorithms (i.e. be a new version rather than a new algorithm name). This algorithm is completely general, it would work for any detector geometry given a correct and calibrated IDF (which we don't have for the old POLDI geometry).

The whole POLDI workflow is being updated, for now any systems tests using old data have been altered to run v5 of   PoldiAutoCorrelation. Once the whole workflow is supported they will be altered to use new data and the new algorithm versions.

Fixes #38898

**Report to:** Florencia Malamud (POLDI, PSI)

### To test:

(1) Run this script

```
from mantid.simpleapi import *
import matplotlib.pyplot as plt
from mantid.api import FileFinder
from plugins.algorithms.poldi_utils import load_poldi, simulate_2d_data
from mantid.geometry import CrystalStructure, ReflectionGenerator, ReflectionConditionFilter

fpath_data = FileFinder.getFullPath("poldi_448x500_chopper5k_silicon.txt")
fpath_idf = FileFinder.getFullPath("POLDI_Definition_448_calibrated.xml")

# load the raw data
ws = load_poldi(fpath_data, fpath_idf, chopper_speed=5000, t0=5.855e-02, t0_const=-9.00)
ws_corr = PoldiAutoCorrelation(InputWorkspace=ws, OutputWorkspace='ws_corr')

# simulate Si workspace
dmax = 2*np.pi/ws_corr.readX(0)[0]
dmin = 2*np.pi/ws_corr.readX(0)[-1]
xtal = CrystalStructure("5.431 5.431 5.431", "F d -3 m", "Si 0 0 0 1.0 0.05")
generator = ReflectionGenerator(xtal, ReflectionConditionFilter.StructureFactor)
hkls = generator.getUniqueHKLs(dmin, dmax)
dlist = generator.getDValues(hkls)
qlist = 2*np.pi/np.array(dlist)

func = FlatBackground(A0=10)
for q in qlist:
    func = func + Gaussian(Height=100, PeakCentre=q, Sigma=0.0015*q)
ws_spec = CreateWorkspace(DataX=ws_corr.readX(0), DataY=func(ws_corr.readX(0)),
                          UnitX="MomentumTransfer", YUnitLabel="Intensity (a.u.)")
                                                   
ws_sim = simulate_2d_data(ws, ws_spec, output_workspace="ws_sim")
ws_sim_corr = PoldiAutoCorrelation(InputWorkspace=ws_sim)

fig, axes = plt.subplots(3, 1, sharex=True, figsize=(6,7), subplot_kw={'projection': 'mantid'})
labels = ["Measured correlation", "Simulated correlation", "Input Spectrum"]
for iax, w in enumerate([ws_corr, ws_sim_corr, ws_spec]):
    axes[iax].plot(w, color='k', ls='-', marker='o', markersize=3, label=labels[iax], wkspIndex=0, distribution=True)
    axes[iax].set_title(labels[iax])
    axes[iax].legend(fontsize=8)
    for q in qlist:
        axes[iax].axvline(q, color='r', ls='--', alpha=0.5, label='Bragg peaks')
fig.tight_layout()
fig.show()
```

It should look like this
![image](https://github.com/user-attachments/assets/c6149d3e-6571-4815-9510-a3eab358a0bd)

The peaks centres in each of the spectra plotted should agree and be consistent with the predicted Bragg reflections (red lines).

Note that the intensities may differ as input spectrum is not realistic (not taking into account structure factor or multiplicity) and the 2D data have not not been normalised by flux etc.


<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

RichardWaiteSTFC committed:
Add v6 POLDIAutoCorrelation algorithm and util functions

RichardWaiteSTFC committed:
Speed up calculation by replacing np.mod

RichardWaiteSTFC committed:
Add input validation to workspace

RichardWaiteSTFC committed:
Convert output to Q to match v5 of POLDIAutoCorrelation

sf1919 commented:
I re-ran the Linux tests (unit & system, packaging and docs) after `main` was fixed yesterday. Please treat those failures as genuine.

RichardWaiteSTFC committed:
Use v5 of PoldiAutoCorrelation in systems test and workflow algorithm

RichardWaiteSTFC committed:
Add function to simulate 2d workspace

RichardWaiteSTFC committed:
Fix out by 1/2 time bin bug in auto-corr

Needed to correct for fact first bin center is at bin_width / 2 not 0

RichardWaiteSTFC committed:
Convert to point data if converting to d-spacing prior to interp

Also remove use of deprecated scipy.interpolate.interp1d

RichardWaiteSTFC committed:
Get max time bin from cycle period (based on chopper speed)

RichardWaiteSTFC committed:
Make t0 parameter correspond to fraction of cycle time

So same parameter can be used for any chopper speed

RichardWaiteSTFC committed:
Add new algorithm doc and test data

RichardWaiteSTFC committed:
Add unit test

sf1919 commented:
I'm not convinced all of the test failures from commit `61484dd` are necessarily your code. Several of them were failres on the nightly too and I think once thats fixed it might fix some tests for you too. Just concentrate on fixing any ones obviously related to this PR.

RichardWaiteSTFC committed:
Fix doc test indent typo on figure caption

RichardWaiteSTFC committed:
Add comment on d-spacing range calculated

RichardWaiteSTFC committed:
Fix doc tests using old POLDI data (prior to upgrade)

This requires calling PoldiAutoCorrelation with Version=5

RichardWaiteSTFC committed:
Fix AlgorithmWorkspaceGroupValidateInputsTest

ws = self.getProperty("InputWorkspace").value returns None is passed a group workspace.
Could have moved these checks into exec but better to tell user in algorithm GUI than in runtime error.
If user does pass a group workspace of invalid workspaces Mantid won't crash the algorithm will just error because it can't find the right component.

RichardWaiteSTFC committed:
Add release notes

And update copyright date in headers

RichardWaiteSTFC commented:
This is a comparison of the correlation spectrum of Si using Version=5 and Version=6 (note different data taken with different detector geometries and calibration (both measured with chopper speed 10000 rpm).
![image](https://github.com/user-attachments/assets/d88580e4-998d-411e-a811-36f71fb75d66)

They show good agreement (note constant background subtracted and scaled to have max = 1)

andy-bridger commented on Framework/PythonInterface/plugins/algorithms/poldi_utils.py:
```diff
@@ -0,0 +1,184 @@
+    :param fpath_data: filepath of ASCII file with counts
+    :param fpath_idf: filepath to instrument definition file
+    :param chopper_speed: chopper speed used on instrument (rpm)
+    :param t0: time offset to applied to slit openings (fraction of cycle time)
```
typo 
```suggestion
    :param t0: time offset applied to slit openings (fraction of cycle time)
```

andy-bridger commented on Framework/PythonInterface/plugins/algorithms/poldi_utils.py:
```diff
@@ -0,0 +1,184 @@
+    cycle_time = _calc_cycle_time_from_chopper_speed(chopper_speed)  # mus
+    # get chopper offsets in time
+    nslits = chopper.nelements()
+    slit_offsets = np.array([(chopper[islit].getPos()[0] + t0) * cycle_time for islit in range(nslits)])
```
Worth adding a comment noting the make up of the chopper position V3D? (is it (slit location as fraction of revolution, 0, z coordinate)?) 

Feel free to ignore this if it is already widely known

RichardWaiteSTFC replied:
I think that's a fair - I will add a comment in the code.
In the IDF it looks like this
```
<type name="sample" is="SamplePos" />
<type name="chopper" is="ChopperPos">
   <component type="chopperSlit">
     <location x="0.000000" name="slit0" />
     <location x="0.162156" name="slit1" />
     <location x="0.250867" name="slit2" />
     <location x="0.3704" name="slit3" />
     <location x="0.439811" name="slit4" />
     <location x="0.588455" name="slit5" />
     <location x="0.761389" name="slit6" />
     <location x="0.895667" name="slit7" />
   </component>
</type>
```
I think when an IDF gets added in the instrument directory it should have a comment in there as well.

andy-bridger commented on Framework/PythonInterface/plugins/algorithms/poldi_utils.py:
```diff
@@ -0,0 +1,184 @@
+
+def get_final_dspac_array(bin_width: float, dspac_min: float, dspac_max: float, time_max: float) -> np.ndarray[float]:
+    """
+    Function to d-spacing bins given time bin width and maximum arrival time
```
typo
```suggestion
    Function to calculate d-spacing bins given time bin width and maximum arrival time
```

andy-bridger commented on Framework/PythonInterface/plugins/algorithms/poldi_utils.py:
```diff
@@ -0,0 +1,184 @@
+def simulate_2d_data(ws_2d: Workspace2D, ws_1d: Workspace2D, output_workspace: Optional[str] = None, lambda_max: float = 5.0):
+    """
+    Function to simulate 2D pulse overlap data given 1D spectrum in d-spacing (i.e. powder pattern).
+    :param ws_2d: w
```
some of this is missing?

RichardWaiteSTFC commented on Framework/PythonInterface/plugins/algorithms/poldi_utils.py:
```diff
@@ -0,0 +1,184 @@
+    return time_max
+
+
+def simulate_2d_data(ws_2d: Workspace2D, ws_1d: Workspace2D, output_workspace: Optional[str] = None, lambda_max: float = 5.0):
```
> Maybe some tests might be useful on simulate_2d_data?

Normally I would agree, in this case it was just a helper function for me and POLDI scientists to test the auto-correlation is performing as expected. The 2D fitting and simulation will eventually be it's own algorithm covered in issue #38970 - as part of which I will add unit tests.

> I was wondering whether the logic on L156 works, if ConvertUnits also gets applied in place when OutputWorkspace is specified as __temp (it seems it must, but not something I have come across before now).

I think you are referring to this line
`ws_1d = exec_alg("ConvertUnits", InputWorkspace=ws_1d, Target="dSpacing", OutputWorkspace="__temp")`
This operation is not being done in-place, the workspace produces will appear in the ADS as "__temp" which by default the user will not be able to see. But the python variable `ws_1d` will now point to this workspace not the workspace passed in (the only case of concern is if the input `ws_1d` workspace is called `__temp`).

Ultimately this is not ideal, one would like the workspace `__temp` not to live in the ADS at all (i.e. call the algorithm  `StoreInADS=False`) - however I remember having issues getting the `exec_alg` function to work. I'll have another go!

andy-bridger replied:
I defer to you in terms of test coverage, perhaps if this is only a temporary function which will be replace with an algorithm that could be noted in a comment/doc string? (might be useful if any users get spurious results?). Also, thanks for the explanation of how `__temp` works!

andy-bridger commented on Framework/PythonInterface/plugins/algorithms/PoldiAutoCorrelation6.py:
```diff
@@ -0,0 +1,136 @@
+        if ws is None:
+            # group workspace
+            return issues
+        if ws.getNumberHistograms() == 400:
```
Is it possible this could get caught if the detector happens to be unfortunately masked?

RichardWaiteSTFC replied:
`getNumberHistograms()` includes masked and monitor spectra - I just wanted to give the user some indication they should use `Version=5` . The algorithm won't crash Mantid etc. if you run it with an old workspace, it's just the result won't make sense as the IDF that ships with Mantid is inaccurate.

RichardWaiteSTFC committed:
Fix doc strings, typos and add comments after review

RichardWaiteSTFC committed:
Use if TYPE_CHECKING for compatibility with future static analysis

See discussion on #38773
https://github.com/mantidproject/mantid/pull/38773/files#r1979127157

RichardWaiteSTFC committed:
Use setAlwaysStoreInADS(False) in algorithm manager

RichardWaiteSTFC commented:
Thanks for the review @andy-bridger - think I got everything...I also change the type hinting in light of discussion here
https://github.com/mantidproject/mantid/pull/38773/files#r1979127157

andy-bridger commented on docs/source/algorithms/PoldiAutoCorrelation-v6.rst:
```diff
@@ -0,0 +1,81 @@
+
+Note by convention the correlation spectrum is converted from d-spacing into momentum transfer.
+
+Further details of the POLDI instrument and the reduction can be found in [1]_.
```
I do have one more thing, but I am happy to do this myself as a separate pull request?

I don't think this ref does a very good job of explaining why any of the steps are being done and what they achieve. Perhaps it is just me, but it took me ages to get my head around the whole concept and I think it could be worth having an extra explainer which doesn't have the same formalisation as an academic paper? 


My proposed explanation is: 
Calculating the Intermediate Correlation Spectrum, conceptually:

As POLDI has overlapped pulses, it is not possible to tell, just from the time of flight and the detector position, what the momentum transfer vector is which gave rise to the event. In a generic system, for a single recorded neutron event, there is a continuous range of momentum transfer vectors which could result in the specific TOF

As the TOF is inversely proportional to the d-spacing, and the total experiment time is fixed, we have can reduce this to a fixed range of d-spacings that could have given rise to any detected event. 
Further, because we have discrete time bins, we can use these to describe a range of all resolvable d-spacings that might have been recorded during the experiment.

The key to overlapped pulses working is that because there is a chopper with known slit positions spinning at a known speed, there is a known offset in time between when one slit allows the beam through, and all the subsequent slits allowing the beam through. This means we know that for any momentum transfer vector (and thus TOF, for a given detector) to actually be experimentally present we must also observe neutron detection with TOFs corresponding to the slit offsets.

This is the key to the Auto Correlation, finding which d-spacings give consistent intensities at the points in time corresponding to the known slits in the chopper.

How this is done:

For the recorded experimental data, for each possible d-spacing, for each detector, there is a specific sequence of points in time that neutrons should be detected, if that d-spacing is present. As we know exactly what the sequence of points in time should be, we can evaluate what the experimental intensity at these points and check that it is consistent across all detectors. As the data is histogram data, evaluating the specific points in time is done by taking a linear combination of the intensity of the bin below and the intensity of the bin above, weighted by how far along the interval the point falls.

Once this has been done, for all possible d-spacings and all detectors, the detector signals can be summed to give the total intensity for each d-spacing at each point in time corresponding to an opening in the chopper.

As only the d-spacings with intensities consistent across all these chopper slits can actually be experimentally present, we require a correlation value which should only be high when all the intensities are high and should be low if any are low. This is achieved by taking the inverse of these summed-across-detector intensities, summing them across all the points in time corresponding to slit openings, and then taking the inverse again. This works because if any of the summed-across-detector intensities are low, their inverse will be large. This will dominate the subsequent sum and when it is then re-inverted, the resulting value will be small. Only when all the summed-across-detector intensities are high, will their inverses be small, and the sum of the inverses also be small, yielding a large final correlation.

RichardWaiteSTFC replied:
I'm not averse to adding some more explanation of how POLDI works though it isn't strictly our remit - for example I don't think we have any documentation on how a conventional TOF diffractometer?

I think if you want to add some more info it is best placed in a concepts document rather than the documentation of the auto-correlation algorithm - but I'm open to discussion.

There are some issues with the explanation above, for example:
1. The TOF is proportional to the d-spacing (not inversely proportional - I think this is just a typo and you meant |Q|)
2. TOF itself is not measured - the x-axis is the time since the last index pulse (the clock is reset at the start of each chopper rotation/period). The neutron could have come from many possible previous pulses (governed by wavelength range and path length from chopper to detector). .
3. I would describe it as an interpolation rather than a 'weighted combination of the intensity of the bin below and the intensity of the bin above [paraphrased]'

I could add more detail about the sum of the inverse - in the docs I describe this as " a type of average of the intensity over different arrival times from each pulse corresponding to a given d-spacing for each detector."  which it is. However this is explained quite well in section 3.1 of the paper (specifically Eq. 8). If you compare what I have to the previous docs - I think you can see I made some effort to explain the basics!
https://docs.mantidproject.org/nightly/algorithms/PoldiAutoCorrelation-v5.html

andy-bridger replied:
I agree that the new docs are much better than before and this isn't really the point for the algorithm docs, so happy to approve here.

We can discuss perhaps adding and linking a concepts discussion at a later date, with my myriad of errors ironed out - I personally didn't find section 3.1 of the paper particularly helpful as to why anything was happening, but agree that it is clear about what the steps are (I guess if you are more familiar with these things than I am, that is probably plenty)

RichardWaiteSTFC committed:
Add more info about the correlation summation/average

RichardWaiteSTFC commented:
Sorry for dismissing your review, I spotted a missing  `ADS.addOrReplace` in `poldi_utils.simulate_2d_data`.

Since I was at it, I also added a bit more info about the summation to calculate the correlation intensity

andy-bridger reviewed (approved):
Looks good and the changes to the docs are helpful, thanks!
